src/machine/machine-varlink.c

```diff
@@ -0,0 +1,171 @@
+                if (b > INT_MAX || b <= 0)
+                        return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "Invalid network interface index %"PRIu64, b);
+
+                netif[k++] = (int) b;
+        }
+        assert(k == n_netif);
+
+        m->netif = TAKE_PTR(netif);
+        m->n_netif = n_netif;
+
+        return 0;
+}
+
+int vl_method_register(Varlink *link, JsonVariant *parameters, VarlinkMethodFlags flags, void *userdata) {
+        Manager *manager = ASSERT_PTR(userdata);
+        _cleanup_(machine_freep) Machine *user_machine = NULL;
+        Machine *managed_machine;
+        int r;
+
+        static const JsonDispatch dispatch_table[] = {
+                { "name",              JSON_VARIANT_STRING,   machine_name,         offsetof(Machine, name),                 JSON_MANDATORY },
+                { "id",                JSON_VARIANT_STRING,   json_dispatch_id128,  offsetof(Machine, id),                   0 },
+                { "service",           JSON_VARIANT_STRING,   json_dispatch_string, offsetof(Machine, service),              0 },
+                { "class",             JSON_VARIANT_STRING,   machine_class,        offsetof(Machine, class),                JSON_MANDATORY },
+                { "leader",            JSON_VARIANT_UNSIGNED, machine_leader,       offsetof(Machine, leader),               0 },
+                { "rootDirectory",     JSON_VARIANT_STRING,   json_dispatch_path,   offsetof(Machine, root_directory),       0 },
+                { "ifIndices",         JSON_VARIANT_ARRAY,    machine_ifindices,    0,                                       0 },
+                { "sshAddress",        JSON_VARIANT_STRING,   json_dispatch_string, offsetof(Machine, ssh_address),          JSON_SAFE },
+                { "sshPrivateKeyPath", JSON_VARIANT_STRING,   json_dispatch_path,   offsetof(Machine, ssh_private_key_path), JSON_SAFE },
+                {}
+        };
+
+        user_machine = new0(Machine, 1);
+        if (!user_machine)
+                return log_oom();
+
+        r = varlink_dispatch(link, parameters, dispatch_table, user_machine);
+        if (r != 0)
+                return r;
+
+        r = manager_add_machine(manager, user_machine->name, &managed_machine);
+        if (r < 0)
+                return r;
+
+        managed_machine->name = user_machine->name;
+        managed_machine->id = user_machine->id;
+        managed_machine->service = user_machine->service;
+        managed_machine->class = user_machine->class;
+        managed_machine->leader = user_machine->leader;
+        managed_machine->root_directory = user_machine->root_directory;
+        managed_machine->netif = user_machine->netif;
+        managed_machine->n_netif = user_machine->n_netif;
+        managed_machine->ssh_address = user_machine->ssh_address;
+        managed_machine->ssh_private_key_path = user_machine->ssh_private_key_path;
+
+        user_machine = mfree(user_machine);
```

Comment:
hmm, i am not a fan of having two different lifecycles for the Machine structure. i.e. once here which is just freed with mfree(), and one via machine_free().

I'd prefer if we just have machine_free() and that's it.

Hence, my proposal would be to split up the existing machine_new() in two: a machine_new() that just allocates the structured, and does basic initialization of some fields, but does not add to the Manager object. And machine_link() or so which actually links it to the Manager object. And machine_free() should be able to deal with linked and unlinked object, i.e. check if the "machine" field is initialized, and if so, it assumes the thing is linked in, otherwise not.

Then we would use machine_new() here, and then use varlink_dispatch() to initialize it, and then link it in once that completed.